[Failing Test]: beam_PerformanceTests_SparkReceiver_IO failing

### What happened?

Since Apr 25, 2023.

Error massage:

```
23:05:17 org.apache.beam.sdk.io.sparkreceiver.SparkReceiverIOIT > testSparkReceiverIOReadsInStreamingWithOffset FAILED
23:05:17     java.lang.AssertionError: expected:<5000000> but was:<4700000>
23:05:17         at org.junit.Assert.fail(Assert.java:89)
23:05:17         at org.junit.Assert.failNotEquals(Assert.java:835)
23:05:17         at org.junit.Assert.assertEquals(Assert.java:647)
23:05:17         at org.junit.Assert.assertEquals(Assert.java:633)
23:05:17         at org.apache.beam.sdk.io.sparkreceiver.SparkReceiverIOIT.testSparkReceiverIOReadsInStreamingWithOffset(SparkReceiverIOI
```

the actual number of element is fluctuating and always smaller than the expected. May indicates a data loss bug

### Issue Failure

Failure: Test is continually failing

### Issue Priority

Priority: 1 (unhealthy code / failing or flaky postcommit so we cannot be sure the product is healthy)

### Issue Components

- [ ] Component: Python SDK
- [X] Component: Java SDK
- [ ] Component: Go SDK
- [ ] Component: Typescript SDK
- [X] Component: IO connector
- [ ] Component: Beam examples
- [ ] Component: Beam playground
- [ ] Component: Beam katas
- [ ] Component: Website
- [ ] Component: Spark Runner
- [ ] Component: Flink Runner
- [ ] Component: Samza Runner
- [ ] Component: Twister2 Runner
- [ ] Component: Hazelcast Jet Runner
- [ ] Component: Google Cloud Dataflow Runner

--- a/sdks/java/io/cdap/src/main/java/org/apache/beam/sdk/io/cdap/CdapIO.java
+++ b/sdks/java/io/cdap/src/main/java/org/apache/beam/sdk/io/cdap/CdapIO.java
@@ -152,7 +152,8 @@ import org.checkerframework.checker.nullness.qual.Nullable;
  *
  * <p>Optionally you can pass {@code pullFrequencySec} which is a delay in seconds between polling
  * for new records updates, you can pass {@code startOffset} which is inclusive start offset from
- * which the reading should be started.
+ * which the reading should be started. Also, you can pass {@code startPollTimeoutSec} which is
+ * delay in seconds before start polling.
  *
  * <p>{@link Plugin} is the Wrapper class for the Cdap Plugin. It contains main information about
  * the Plugin. The object of the {@link Plugin} class can be created with the {@link
@@ -186,6 +187,7 @@ import org.checkerframework.checker.nullness.qual.Nullable;
  *             .withKeyClass(String.class)
  *             .withValueClass(String.class)
  *             .withPullFrequencySec(1L)
+ *             .withStartPollTimeoutSec(2L)
  *             .withStartOffset(10L);
  * }</pre>
  */
@@ -226,6 +228,8 @@ public class CdapIO {
 
     abstract @Nullable Long getPullFrequencySec();
 
+    abstract @Nullable Long getStartPollTimeoutSec();
+
     abstract @Nullable Long getStartOffset();
 
     abstract Builder<K, V> toBuilder();
@@ -242,6 +246,8 @@ public class CdapIO {
       abstract Builder<K, V> setValueClass(Class<V> valueClass);
 
       abstract Builder<K, V> setPullFrequencySec(Long pullFrequencySec);
+
+      abstract Builder<K, V> setStartPollTimeoutSec(Long startPollTimeoutSec);
 
       abstract Builder<K, V> setStartOffset(Long startOffset);
 
@@ -288,6 +294,12 @@ public class CdapIO {
       return toBuilder().setPullFrequencySec(pullFrequencySec).build();
     }
 
+    /** Delay in seconds before start polling. Applicable only for streaming Cdap Plugins. */
+    public Read<K, V> withStartPollTimeoutSec(Long startPollTimeoutSec) {
+      checkArgument(startPollTimeoutSec != null, "Start poll timeout can not be null");
+      return toBuilder().setStartPollTimeoutSec(startPollTimeoutSec).build();
+    }
+
     /**
      * Inclusive start offset from which the reading should be started. Applicable only for
      * streaming Cdap Plugins.
@@ -326,6 +338,10 @@ public class CdapIO {
         Long pullFrequencySec = getPullFrequencySec();
         if (pullFrequencySec != null) {
           reader = reader.withPullFrequencySec(pullFrequencySec);
+        }
+        Long startPollTimeoutSec = getStartPollTimeoutSec();
+        if (startPollTimeoutSec != null) {
+          reader = reader.withStartPollTimeoutSec(startPollTimeoutSec);
         }
         Long startOffset = getStartOffset();
         if (startOffset != null) {

--- a/sdks/java/io/sparkreceiver/2/src/main/java/org/apache/beam/sdk/io/sparkreceiver/ReadFromSparkReceiverWithOffsetDoFn.java
+++ b/sdks/java/io/sparkreceiver/2/src/main/java/org/apache/beam/sdk/io/sparkreceiver/ReadFromSparkReceiverWithOffsetDoFn.java
@@ -62,8 +62,8 @@ class ReadFromSparkReceiverWithOffsetDoFn<V> extends DoFn<byte[], V> {
   private static final Logger LOG =
       LoggerFactory.getLogger(ReadFromSparkReceiverWithOffsetDoFn.class);
 
-  /** Constant waiting time after the {@link Receiver} starts. Required to prepare for polling */
-  private static final long START_POLL_TIMEOUT_MS = 2000;
+  /** Waiting time after the {@link Receiver} starts. Required to prepare for polling */
+  private static final long DEFAULT_START_POLL_TIMEOUT_SEC = 2;
 
   /** Delay between polling for new records updates. */
   private static final long DEFAULT_PULL_FREQUENCY_SEC = 0;
@@ -77,6 +77,7 @@ class ReadFromSparkReceiverWithOffsetDoFn<V> extends DoFn<byte[], V> {
   private final SerializableFunction<V, Instant> getTimestampFn;
   private final ReceiverBuilder<V, ? extends Receiver<V>> sparkReceiverBuilder;
   private final Long pullFrequencySec;
+  private final Long startPollTimeoutSec;
   private final Long startOffset;
 
   ReadFromSparkReceiverWithOffsetDoFn(SparkReceiverIO.Read<V> transform) {
@@ -102,6 +103,12 @@ class ReadFromSparkReceiverWithOffsetDoFn<V> extends DoFn<byte[], V> {
       pullFrequencySec = DEFAULT_PULL_FREQUENCY_SEC;
     }
     this.pullFrequencySec = pullFrequencySec;
+
+    Long startPollTimeoutSec = transform.getStartPollTimeoutSec();
+    if (startPollTimeoutSec == null) {
+      startPollTimeoutSec = DEFAULT_START_POLL_TIMEOUT_SEC;
+    }
+    this.startPollTimeoutSec = startPollTimeoutSec;
 
     Long startOffset = transform.getStartOffset();
     if (startOffset == null) {
@@ -147,7 +154,7 @@ class ReadFromSparkReceiverWithOffsetDoFn<V> extends DoFn<byte[], V> {
 
         if (split != null) {
           OffsetRange primary = split.getPrimary();
-          if (primary != null && primary.getFrom() == primary.getTo() && !isCheckDoneCalled.get()) {
+          if (primary != null && !isCheckDoneCalled.get()) {
             // If there was no check done called, then the split is not needed
             LOG.debug("Split is not needed");
             isCheckDoneCalled.set(false);
@@ -283,7 +290,7 @@ class ReadFromSparkReceiverWithOffsetDoFn<V> extends DoFn<byte[], V> {
     while (true) {
       LOG.debug("Start polling records");
       try {
-        TimeUnit.MILLISECONDS.sleep(START_POLL_TIMEOUT_MS);
+        TimeUnit.SECONDS.sleep(startPollTimeoutSec);
       } catch (InterruptedException e) {
         LOG.error("SparkReceiver was interrupted before polling started", e);
         throw new IllegalStateException("Spark Receiver was interrupted before polling started");

--- a/sdks/java/io/sparkreceiver/2/src/main/java/org/apache/beam/sdk/io/sparkreceiver/SparkReceiverIO.java
+++ b/sdks/java/io/sparkreceiver/2/src/main/java/org/apache/beam/sdk/io/sparkreceiver/SparkReceiverIO.java
@@ -50,7 +50,8 @@ import org.slf4j.LoggerFactory;
  * startOffset} which is inclusive start offset from which the reading should be started.
  *
  * <p>Optionally you can pass {@code pullFrequencySec} which is a delay in seconds between polling
- * for new records updates.
+ * for new records updates. Also, you can pass {@code startPollTimeoutSec} which is delay in seconds
+ * before start polling.
  *
  * <p>Example of {@link SparkReceiverIO#read()} usage:
  *
@@ -68,6 +69,7 @@ import org.slf4j.LoggerFactory;
  *      .withGetOffsetFn(Long::valueOf)
  *      .withTimestampFn(Instant::parse)
  *      .withPullFrequencySec(1L)
+ *      .withStartPollTimeoutSec(2L)
  *      .withStartOffset(10L)
  *      .withSparkReceiverBuilder(receiverBuilder);
  * }</pre>
@@ -93,6 +95,8 @@ public class SparkReceiverIO {
 
     abstract @Nullable Long getPullFrequencySec();
 
+    abstract @Nullable Long getStartPollTimeoutSec();
+
     abstract @Nullable Long getStartOffset();
 
     abstract Builder<V> toBuilder();
@@ -108,6 +112,8 @@ public class SparkReceiverIO {
       abstract Builder<V> setTimestampFn(SerializableFunction<V, Instant> timestampFn);
 
       abstract Builder<V> setPullFrequencySec(Long pullFrequencySec);
+
+      abstract Builder<V> setStartPollTimeoutSec(Long startPollTimeoutSec);
 
       abstract Builder<V> setStartOffset(Long startOffset);
 
@@ -137,6 +143,12 @@ public class SparkReceiverIO {
     public Read<V> withPullFrequencySec(Long pullFrequencySec) {
       checkArgument(pullFrequencySec != null, "Pull frequency can not be null");
       return toBuilder().setPullFrequencySec(pullFrequencySec).build();
+    }
+
+    /** Waiting time after the {@link Receiver} starts. Required to prepare for polling. */
+    public Read<V> withStartPollTimeoutSec(Long startPollTimeoutSec) {
+      checkArgument(startPollTimeoutSec != null, "Start poll timeout can not be null");
+      return toBuilder().setStartPollTimeoutSec(startPollTimeoutSec).build();
     }
 
     /** Inclusive start offset from which the reading should be started. */

--- a/sdks/java/io/sparkreceiver/2/src/test/java/org/apache/beam/sdk/io/sparkreceiver/ReadFromSparkReceiverWithOffsetDoFnTest.java
+++ b/sdks/java/io/sparkreceiver/2/src/test/java/org/apache/beam/sdk/io/sparkreceiver/ReadFromSparkReceiverWithOffsetDoFnTest.java
@@ -119,13 +119,14 @@ public class ReadFromSparkReceiverWithOffsetDoFnTest {
             TEST_ELEMENT, dofnInstance.initialRestriction(TEST_ELEMENT));
 
     assertTrue(offsetRangeTracker.tryClaim(0L));
+    // Split is not needed, because check done wasn't called
+    assertNull(offsetRangeTracker.trySplit(0d));
+
+    offsetRangeTracker.checkDone();
+    // Perform split because check done was called
     assertEquals(
         SplitResult.of(new OffsetRange(0, 1), new OffsetRange(1, Long.MAX_VALUE)),
         offsetRangeTracker.trySplit(0d));
-
-    offsetRangeTracker.checkDone();
-    // Split is not needed
-    assertNull(offsetRangeTracker.trySplit(0d));
   }
 
   @Test

--- a/sdks/java/io/sparkreceiver/2/src/test/java/org/apache/beam/sdk/io/sparkreceiver/SparkReceiverIOTest.java
+++ b/sdks/java/io/sparkreceiver/2/src/test/java/org/apache/beam/sdk/io/sparkreceiver/SparkReceiverIOTest.java
@@ -42,6 +42,7 @@ public class SparkReceiverIOTest {
   public static final TestPipelineOptions OPTIONS =
       TestPipeline.testingPipelineOptions().as(TestPipelineOptions.class);
   public static final long PULL_FREQUENCY_SEC = 1L;
+  public static final long START_POLL_TIMEOUT_SEC = 2L;
   public static final long START_OFFSET = 0L;
 
   static {
@@ -62,6 +63,7 @@ public class SparkReceiverIOTest {
             .withGetOffsetFn(offsetFn)
             .withTimestampFn(timestampFn)
             .withPullFrequencySec(PULL_FREQUENCY_SEC)
+            .withStartPollTimeoutSec(START_POLL_TIMEOUT_SEC)
             .withStartOffset(START_OFFSET)
             .withSparkReceiverBuilder(receiverBuilder);
 
@@ -96,6 +98,13 @@ public class SparkReceiverIOTest {
   }
 
   @Test
+  public void testReadObjectCreationFailsIfStartPollTimeoutSecIsNull() {
+    assertThrows(
+        IllegalArgumentException.class,
+        () -> SparkReceiverIO.<String>read().withStartPollTimeoutSec(null));
+  }
+
+  @Test
   public void testReadObjectCreationFailsIfStartOffsetIsNull() {
     assertThrows(
         IllegalArgumentException.class, () -> SparkReceiverIO.<String>read().withStartOffset(null));
@@ -126,6 +135,7 @@ public class SparkReceiverIOTest {
             .withGetOffsetFn(Long::valueOf)
             .withTimestampFn(Instant::parse)
             .withPullFrequencySec(PULL_FREQUENCY_SEC)
+            .withStartPollTimeoutSec(START_POLL_TIMEOUT_SEC)
             .withStartOffset(START_OFFSET)
             .withSparkReceiverBuilder(receiverBuilder);
 
@@ -149,6 +159,7 @@ public class SparkReceiverIOTest {
             .withGetOffsetFn(Long::valueOf)
             .withTimestampFn(Instant::parse)
             .withPullFrequencySec(PULL_FREQUENCY_SEC)
+            .withStartPollTimeoutSec(START_POLL_TIMEOUT_SEC)
             .withStartOffset(START_OFFSET)
             .withSparkReceiverBuilder(receiverBuilder);
 
@@ -171,6 +182,7 @@ public class SparkReceiverIOTest {
             .withGetOffsetFn(Long::valueOf)
             .withTimestampFn(Instant::parse)
             .withPullFrequencySec(PULL_FREQUENCY_SEC)
+            .withStartPollTimeoutSec(START_POLL_TIMEOUT_SEC)
             .withStartOffset(START_OFFSET)
             .withSparkReceiverBuilder(receiverBuilder);
 
@@ -193,6 +205,7 @@ public class SparkReceiverIOTest {
             .withGetOffsetFn(Long::valueOf)
             .withTimestampFn(Instant::parse)
             .withPullFrequencySec(PULL_FREQUENCY_SEC)
+            .withStartPollTimeoutSec(START_POLL_TIMEOUT_SEC)
             .withStartOffset(START_OFFSET)
             .withSparkReceiverBuilder(receiverBuilder);
 
@@ -215,6 +228,7 @@ public class SparkReceiverIOTest {
             .withGetOffsetFn(Long::valueOf)
             .withTimestampFn(Instant::parse)
             .withPullFrequencySec(PULL_FREQUENCY_SEC)
+            .withStartPollTimeoutSec(START_POLL_TIMEOUT_SEC)
             .withStartOffset(START_OFFSET)
             .withSparkReceiverBuilder(receiverBuilder);
 

